### PR TITLE
Add AbortController implementation

### PR DESCRIPTION
**Pull request by MozarellaMan**
### Issue # (if available)

https://github.com/awslabs/llrt/issues/248

### Description of changes

Adds an implementation for `AbortController` with stubbed methods for `AbortSignal` which would have the actual event aborting implementation.  This implementation attempts to follow the specification found here https://dom.spec.whatwg.org/#ref-for-dom-abortcontroller-abortcontroller%E2%91%A0

### Checklist

- [x] Created unit tests in `tests/unit` and/or in Rust for my feature if needed
- [x] Ran `make fix` to format JS and apply Clippy auto fixes
- [x] Made sure my code didn't add any additional warnings: `make check`
- [ ] Updated documentation if needed ([API.md](API.md)/[README.md](README.md)/Other)

_By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice._

### CODE REVIEW

File: src/events.rs [hunk 1]
```diff
@@ -372,29 +375,95 @@ impl<'js> EventEmitter<'js> {
...
+            reason: None,
+        }
+    }
+
+    #[qjs(get)]
+    pub fn aborted(&self) -> bool {
+        self.aborted
+    }
+
+    pub fn set_reason(&mut self, reason: Opt<Value<'js>>) -> Result<Option<Value<'js>>> {
```
Comment (richarddavison) [inline src/events.rs:450]: Use `#[qjs(skip)]` or move to another `impl`

Comment (richarddavison) [reply src/events.rs:450]: Should this also set aborted to true?

File: src/events.rs [hunk 2]
```diff
@@ -372,29 +375,95 @@ impl<'js> EventEmitter<'js> {
...
+        let new_reason = mem::replace(&mut self.reason, reason.0);
```
Comment (richarddavison) [inline src/events.rs:451]: Is mem::replace same as `self.reason.replace()`?

**Review by richarddavison — CHANGES_REQUESTED** [review]
Looking great so far, thank you for the contributions! Added some minor comments

File: src/events.rs (hunk 2 above)
Comment (MozarellaMan) [reply src/events.rs:452]: I thought so, but it looks like `Option::replace()` takes `T` not `Option<T>`, so I'd have to unwrap the `reason` parameter first. Then I guess set `self.reason` to None if `reason` is None?

File: src/events.rs (hunk 1 above)
Comment (MozarellaMan) [reply src/events.rs:451]: > Should this also set aborted to true?

I wasn't sure as the spec didn't mention the abort controller setting `aborted` to true. I assume maybe whatever observes that the `reason` has changed would then set `aborted` to true. I could set it to true for now till that is implemented

Comment (MozarellaMan) [reply src/events.rs:451]: Also what does `qjs(skip)` do? I couldn't find any documentation for it 👀

Comment (richarddavison) [reply src/events.rs:451]: `qjs(skip)` will not expose the function to JS user land.

File: src/events.rs (hunk 2 above)
Comment (richarddavison) [reply src/events.rs:452]: there is `Option::swap` also 👀

Comment (MozarellaMan) [reply src/events.rs]: I couldn't find an `Option::swap` 🤔 It feels like it should exist though

Comment (MozarellaMan) [reply src/events.rs]: The test I wrote is still failing on asserting the `reason` when `abort()` is called, but it looks like `aborted` is being set to true at least. Will look into it

File: src/events.rs [hunk 3]
```diff
@@ -372,29 +372,112 @@ impl<'js> EventEmitter<'js> {
...
+        AbortSignal {
+            aborted: true,
+            reason: reason.0,
+        }
+    }
+
+    // TODO: Returns an AbortSignal that will automatically abort after a specified time.
+    #[qjs(static)]
+    pub fn timeout() -> AbortSignal<'js> {
+        todo!()
```
Comment (richarddavison) [inline src/events.rs:480]: Here you can simply use `ctx.spawn_exit` to spawn a task and then set abort after the specified time. First create a new abort signal and convert it to a Value: `AbortSignal{...}.into_js()`. Then clone it and return one copy while pass the other copy to the async task which then sets the abort value after said time. Something like:
```rust
 pub fn timeout<'js>(ctx: Ctx<'js>, mills: size) -> Result<Class<'js,Self>> {
        let signal = AbortSignal {
            aborted: true,
            reason: reason.0,
        }
        
        let signal_instance =  Class::instance(ctx,signal);
        let signal_instance2 = signal_instance.clone()
        
        ctx.spawn_exit(async move {
           tokio::time::sleep(mills).await;
           let borrow = signal_instance.borrow();
           borrow.aborted = true;
           borrow.reason = "Timeout" //check spec
        })?;
        Ok(signal_instance2);
    }

**Review by richarddavison — CHANGES_REQUESTED** [review]

File: src/events.rs (hunk 3 above)
Comment (MozarellaMan) [reply src/events.rs]: Thanks! That's really neat 😮 One question about the spec - it mentions putting `DOMException` objects as the `reason` in different cases. I'm not sure if I can I just use the quickjs `Exception` struct here - it looks like `DOMException` is a specific [interface type](https://webidl.spec.whatwg.org/#idl-DOMException). I could include making that in this PR if necessary

Comment (richarddavison) [reply src/events.rs]: Mark it as a todo and we'll address it later :) DOMEx feels a bit out of scope for this PR

File: src/events.rs [hunk 4]
```diff
@@ -474,10 +481,25 @@ impl<'js> AbortSignal<'js> {
...
+        };
+
+        let signal_instance = Class::instance(ctx.clone(), signal)?;
+        let signal_instance2 = signal_instance.clone();
+
+        ctx.spawn_exit(async move {
+            tokio::time::sleep(Duration::from_millis(milliseconds)).await;
+            let borrow = signal_instance.borrow_mut();
+            borrow.set_aborted(true);
+            borrow.set_reason(Opt(None)); // TODO: TimeoutError DOMException
```
Comment (richarddavison) [inline src/events.rs:498]: Use an an Error class

**Review by richarddavison — CHANGES_REQUESTED** [review]

File: src/events.rs (hunk 3 above)
Comment (MozarellaMan) [reply src/events.rs]: Wasn't sure the best way to test this, doesn't seem practical with `jest`

**Review by richarddavison — APPROVED** [review]

**Pull request merged**